source/common/router/config_impl.cc

```diff
@@ -59,6 +59,27 @@ void mergeTransforms(Http::HeaderTransforms& dest, const Http::HeaderTransforms&
                                 src.headers_to_remove.end());
 }
 
+const envoy::config::route::v3::WeightedCluster::ClusterWeight& validateWeightedClusterSpecifier(
+    const envoy::config::route::v3::WeightedCluster::ClusterWeight& cluster) {
+  if (!cluster.name().empty() && !cluster.cluster_header().empty()) {
+    throw EnvoyException("Only one of name or cluster_header can be specified");
+  } else if (cluster.name().empty() && cluster.cluster_header().empty()) {
+    throw EnvoyException("At least one of name or cluster_header need to be specified");
+  } else {
+    if (!cluster.cluster_header().empty()) {
+      envoy::type::matcher::v3::RegexMatcher matcher;
+      const std::string non_strict_pattern = "^[^\\x{0000}\\x{000A}\\x{000D}]*$";
+      matcher.mutable_google_re2();
+      matcher.set_regex(non_strict_pattern);
+      if (!Regex::Utility::parseRegex(matcher)->match(cluster.cluster_header())) {
```

Comment:
Basic idea here makes sense, but why can't you use a PGV annotation for `cluster_header` here? This is a new field, so we shouldn't have to do the HTTP header validation work manually, we can just use the annotation.

Reply:
Thanks for the review! 
We cannot do the full PGV annotation with `min_len: 1` here because it is not `oneof` field along with `name` yet. But we can just do PGV annotation for HTTP header validation part, which i agree it will be cleaner than manual RE2 validation in the implementation.

Done, I updated the corresponding unit test as well.  PTAL!